File: android/mobileclient/app/src/main/java/com/auditpro/mobile_client/pages/ReviewPage.java
/*
  AuditPRO Mobile Client Android
  Copyright 2018 AuditPRO All Rights Reserved
 */
package com.auditpro.mobile_client.pages;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.FileProvider;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.auditpro.mobile_client.Analytics;
import com.auditpro.mobile_client.MainActivity;
import com.auditpro.mobile_client.MobileClientException;
import com.auditpro.mobile_client.api.ApiClient;
import com.auditpro.mobile_client.database.AuditDatabase;
import com.auditpro.mobile_client.entities.Audit;
import com.auditpro.mobile_client.security.Security;
import com.auditpro.mobile_client.test.R;
import com.crashlytics.android.Crashlytics;

import java.io.File;
import java.io.FileFilter;
import java.io.FileOutputStream;
import java.io.UnsupportedEncodingException;
import java.lang.ref.WeakReference;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;


/**
 * Manages a fragment to display closed audits for review.
 * @author Eric Ruck
 */
public class ReviewPage extends BasePage {

	/**
	 * Required empty public constructor.
	 */
	public ReviewPage() {
		// Required empty public constructor
	}

	/**
	 * Use this factory method to create a new instance of
	 * this fragment.
	 * @return A new instance of fragment ReviewPage.
	 */
	public static ReviewPage newInstance() {
		return new ReviewPage();
	}

	/**
	 * Creates our view.
	 * @param inflater Inflater to use
	 * @param container Parent container for view
	 * @param savedInstanceState Optional state
	 * @return New view for fragment
	 */
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		return inflater.inflate(R.layout.fragment_review_page, container, false);
	}


	/**
	 * Initializes the view after it has been created.
	 * @param view Created view
	 * @param savedInstanceState Optional state
	 */
	@Override
	public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		auditList = view.findViewById(R.id.auditList);

		// Connect to the audits database
		try (AuditDatabase db = new AuditDatabase(getContext())) {
			// Get all the closed audits
			Security sec = new Security(getContext().getApplicationContext());
			completedAudits = db.getCompleteAudits(sec.getUserId());
		}
		if ((completedAudits == null) || (completedAudits.size() == 0)) {
			// No audits
			showNoAudits(view);
		}

		// Display the completed audits
		auditAdapter = new ArrayAdapter<Audit>(getContext(), android.R.layout.simple_list_item_2,
				android.R.id.text1, completedAudits) {
			@NonNull @Override
			public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
				View view = super.getView(position, convertView, parent);
				Audit audit = getItem(position);
				if (audit != null) {
					DateFormat formatter = DateFormat.getDateInstance(DateFormat.SHORT);
					((TextView) view.findViewById(android.R.id.text1)).setText(audit.getStoreDescr());
					((TextView) view.findViewById(android.R.id.text2)).setText(formatter.format(audit.getAuditStartedAt()));
				}
				return view;
			}
		};
		auditList.setAdapter(auditAdapter);
		auditList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
			@Override
			public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
				onSelectAudit(auditAdapter.getItem(i));
			}
		});
	}

	/**
	 * Cancels login task in progress on stop.
	 */
	@Override
	public void onStop() {
		super.onStop();
		SyncTask inProgress = (syncTask != null) ? syncTask.get() : null;
		if (inProgress != null) {
			inProgress.cancel(true);
		}
	}

	/**
	 * Provides a title for the action bar.
	 * @param context Application context
	 * @return Action bar title
	 */
	@Override
	public String getPageName(Context context) {
		return context.getResources().getString(R.string.page_name_review);
	}

	/**
	 * Shows the no audits state.
	 * @param view Our view
	 */
	private void showNoAudits(View view) {
		if (view != null) {
			view.findViewById(R.id.noAuditsText).setVisibility(View.VISIBLE);
			auditList.setVisibility(View.GONE);
		}
		if (syncMenuItem != null) {
			syncMenuItem.setVisible(false);
		}
	}

	/**
	 * Shows generic error dialog for page.
	 * @param message Message to show
	 */
	private void showError(String message) {
		Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
	}

	/**
	 * Shows generic error dialog for page.
	 * @param message Message to show
	 */
	private void showError(int message) {
		Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
	}

	/**
	 * Handles an audit selected from the list.
	 * @param audit Selected audit
	 */
	private void onSelectAudit(final Audit audit) {
		// Validate
		if (audit == null) {
			// Nothing to do
			return;
		}

		// Format audit description
		DateFormat formatter = DateFormat.getDateInstance(DateFormat.SHORT);
		String title = String.format("%s\nStarted at %s\nEnded at %s",
				audit.getStoreDescr(),
				formatter.format(audit.getAuditStartedAt()),
				formatter.format(audit.getAuditEndedAt()));

		// Show selected audit with options
		CharSequence[] items = new CharSequence[] {
				getResources().getString(R.string.message_review_opt_sync),
				getString(R.string.message_review_opt_email),
				getString(R.string.message_review_opt_edit),
				getString(R.string.message_review_opt_remove),
				getString(R.string.button_cancel)
		};
		new AlertDialog.Builder(getContext())
				.setTitle(title)
				.setItems(items, new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialogInterface, int i) {
						dialogInterface.dismiss();
						switch (i) {
							case 0: onSync(audit); break;
							case 1: onEmail(audit); break;
							case 2: onEdit(audit); break;
							case 3: onRemove(audit); break;
						}
					}
				})
				.show();
	}

	/**
	 * Confirms request to sync the passed audit.
	 * @param audit Audit to sync
	 */
	private void onSync(final Audit audit) {
		String message = getString(R.string.message_review_confirm_sync, audit.toString());
		new AlertDialog.Builder(getContext())
				.setMessage(message)
				.setNegativeButton(R.string.button_cancel, new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialogInterface, int i) {
						dialogInterface.dismiss();
						Analytics.log("Review", audit,"Sync", "Cancel");
					}
				})
				.setPositiveButton(R.string.button_yes, new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialogInterface, int i) {
						dialogInterface.dismiss();
						Analytics.log("Review", audit,"Sync", "Sync");
						performSync(audit);
					}
				})
				.show();
	}

	/**
	 * Performs the sync of a single audit.
	 * @param audit Audit to sync
	 */
	private void performSync(final Audit audit) {
		syncTask = new WeakReference<>(new SyncTask(this, audit));
		syncTask.get().execute();
	}

	/**
	 * Provides reference-safe async task to login on a worker thread.
	 */
	private static class SyncTask extends AsyncTask<Void, Void, String> {
		private WeakReference<ReviewPage> host;
		private Audit audit;
		private String token;
		private ApiClient api;

		/**
		 * Initialize task to call the remote log in web service.
		 * @param host Parent host page
		 */
		SyncTask(ReviewPage host, Audit audit) {
			this.host = new WeakReference<>(host);
			this.audit = audit;
			this.token = ((MainActivity) host.getActivity()).getSessionToken();
		}

		/**
		 * Shows activity spinner while we run.
		 */
		@Override
		protected void onPreExecute() {
			ReviewPage self = host.get();
			IPageParent parent = (self == null) ? null : self.getParent();
			if (parent != null) {
				parent.setActivity(true);
			}
		}

		/**
		 * Execute remove login web service
		 * @param voids Placeholder for no parameters
		 * @return Error message or null on success
		 */
		@Override
		protected String doInBackground(Void... voids) {
			api = new ApiClient(token);
			try {
				// Send pending audits
				Context context = host.get().getContext();
				try (AuditDatabase db = new AuditDatabase(context)) {
					// Post the current audit
					String json = db.serializeAudit(audit);
					if (!api.postPayload(json)) {
						// Failed to post the audit
						throw new MobileClientException(api.getMessage());
					}

					// Remove the posted audit from our cache
					db.deleteAudit(audit);
				}

				// Indicate success
				return null;
			}
			catch (MobileClientException exc) {
				// Failed to sync
				return exc.getMessage();
			}
		}

		/**
		 * Handles the web service result in the user interface.
		 * @param errorMessage Error message on failure or null on success
		 */
		@Override
		protected void onPostExecute(String errorMessage) {
			super.onPostExecute(errorMessage);
			ReviewPage self = host.get();
			IPageParent parent = (self == null) ? null : self.getParent();
			if (parent == null) {
				// The page was popped before we completed
				return;
			}

			// Update the UI for a completed audit sync
			parent.setActivity(false);
			self.auditAdapter.remove(audit);
			if (self.auditAdapter.getCount() == 0) {
				// No more audits
				self.showNoAudits(self.getView());
			}
			if (errorMessage != null) {
				// Failed to sync, show error toast
				self.showError(errorMessage);
			}
		}
	}

	/**
	 * Displays our menu when we have unscanned items.
	 * @param inflater Inflater to use on menu resource
	 * @param menu Parent receives menu items
	 * @return Handled flag
	 */
	@Override
	public boolean onCreateMenu(MenuInflater inflater, Menu menu) {
		if (completedAudits.size() > 0) {
			// Show the menu
			inflater.inflate(R.menu.menu_review, menu);
			syncMenuItem = menu.findItem(R.id.action_sync_all);
			return true;
		}

		// No audits means no menu
		return false;
	}

	/**
	 * Handles our menu items when they are selected
	 * @param item Item selected
	 * @return Handled flag
	 */
	@Override
	public boolean onMenuItem(MenuItem item) {
		// Are we interested in this item?
		int id = item.getItemId();
		if (id == R.id.action_sync_all) {
			// Offer set unscanned options
			int message = ((MainActivity) getActivity()).isWifiConnected()
					? R.string.message_review_sync_confirm
					: R.string.message_review_sync_confirm_cell;
			new AlertDialog.Builder(getContext())
					.setTitle(R.string.message_review_sync_title)
					.setMessage(message)
					.setPositiveButton(R.string.button_sync, new DialogInterface.OnClickListener() {
						@Override
						public void onClick(DialogInterface dialog, int which) {
							Analytics.menuAction("Sync All", "sync");
							dialog.dismiss();
							IPageParent parent = getParent();
							if (parent != null) {
								parent.requestSync();
							}
						}
					})
					.setNegativeButton(R.string.button_cancel, new DialogInterface.OnClickListener() {
						@Override
						public void onClick(DialogInterface dialog, int which) {
							Analytics.menuAction("Sync All", "cancel");
							dialog.dismiss();
						}
					})
					.show();
			return true;
		}

		// Not handled
		return false;
	}

	/**
	 * Sets up an e-mail for the requested audit.
	 * @param audit Audit to e-mail
	 */
	private void onEmail(Audit audit) {
		// Access the database to serialize the audit
		byte[] payload;
		try (AuditDatabase db = new AuditDatabase(getContext())) {
			// Serialize the audit
			String json = db.serializeAudit(audit);
			payload = json.getBytes("UTF-8");
		} catch (MobileClientException exc) {
			// Failed to serialize audit
			showError(R.string.message_review_error_serialize);
			return;
		} catch (UnsupportedEncodingException exc) {
			// Failed to encode audit
			Crashlytics.log(Log.ERROR, LOG_TAG, String.format("Failed to encode audit %s: %s",
					audit.toString(), exc.getMessage()));
			showError(R.string.message_review_error_encoding);
			return;
		}

		// Prepare to write attachment to app cache
		File cacheFile = getContext().getCacheDir();

		// Cleanup old attachments
		File[] oldAttachments = cacheFile.listFiles(new FileFilter() {
			@Override
			public boolean accept(File pathname) {
				String fileName = pathname.getName();
				return fileName.startsWith("audit-") && fileName.endsWith(".json");
			}
		});
		for (File oldFile : oldAttachments) {
			if (!oldFile.delete()) {
				Log.w(LOG_TAG, String.format("Failed to delete old attachment file %s", oldFile.getAbsoluteFile()));
			}
		}

		// Write the audit attachment to shared cache
		String stamp = new SimpleDateFormat("yyMMddHHmmssZ", Locale.getDefault()).
				format(audit.getAuditStartedAt());
		File attachmentFile = new File(cacheFile, String.format(
				"audit-%s.json", stamp
		));
		try (FileOutputStream outs = new FileOutputStream(attachmentFile)) {
			outs.write(payload);
		} catch (Exception exc) {
			Log.w(LOG_TAG, "Failed to write completed audit JSON for attachment", exc);
			return;
		}

		// Setup e-mail
		Uri attachmentUri = FileProvider.getUriForFile(getContext(),
				"com.auditpro.mobile_client.fileprovider", attachmentFile);
		ArrayList<Uri> attachments = new ArrayList<>();
		attachments.add(attachmentUri);
		Intent intent = new Intent(Intent.ACTION_SEND_MULTIPLE);
		intent.setType("text/email");
		intent.putExtra(android.content.Intent.EXTRA_EMAIL, new String[] { EMAIL_TO });
		intent.putExtra(android.content.Intent.EXTRA_SUBJECT,"Completed Audit");
		intent.putExtra(android.content.Intent.EXTRA_TEXT,
				String.format("Completed audit of %s attached", audit.toString()));
		intent.putParcelableArrayListExtra(Intent.EXTRA_STREAM, attachments);
		try {
			// Show the configured e-mail client
			startActivity(intent);
			Analytics.log("Review", audit,"Email", "Email");
		} catch (Exception exc) {
			// Unable to show configured e-mail client
			Log.w(LOG_TAG, "Failed to start e-mail activity", exc);
			showError(R.string.message_review_error_email);
		}
	}

	/**
	 * Reopens the passed audit for editing.
	 * @param audit Audit to edit
	 */
	private void onEdit(final Audit audit) {
		try (AuditDatabase db = new AuditDatabase(getContext())) {
			// Is there already an open audit?
			Security sec = new Security(getContext().getApplicationContext());
			if (db.resumeAudit(sec.getUserId()) != null) {
				// Yes, unable to edit now
				Toast.makeText(getContext(), R.string.message_review_error_edit_started, Toast.LENGTH_LONG).show();
				return;
			}

			// Reopen the audit
			Audit reopened = db.reopenAudit(audit);
			IPageParent parent = getParent();
			if (parent != null) {
				Analytics.log("Review", audit, "Edit", "Reopen");
				getParent().swapPage(SelectProductPage.newInstance(reopened));
			}
		} catch (MobileClientException exc) {
			// Show the error to the user, details already logged
			Toast.makeText(getContext(), exc.getMessage(), Toast.LENGTH_SHORT).show();
		}
	}

	/**
	 * Confirms request to remove the passed audit.
	 * @param audit Audit to remove
	 */
	private void onRemove(final Audit audit) {
		String message = getString(R.string.message_review_confirm_remove, audit.toString());
		new AlertDialog.Builder(getContext())
				.setMessage(message)
				.setNegativeButton(R.string.button_cancel, new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialogInterface, int i) {
						dialogInterface.dismiss();
						Analytics.log("Review", audit,"Remove", "Cancel");
					}
				})
				.setPositiveButton(R.string.button_yes, new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialogInterface, int i) {
						dialogInterface.dismiss();
						Analytics.log("Review", audit,"Remove", "Remove");
						performRemove(audit);
					}
				})
				.show();
	}

	/**
	 * Removes the passed audit from the database.
	 * @param audit Audit to remove
	 */
	private void performRemove(final Audit audit) {
		try (AuditDatabase db = new AuditDatabase(getContext())) {
			// Remove the posted audit from our cache
			db.deleteAudit(audit);

			// Remove the audit from the list
			auditAdapter.remove(audit);
			if (auditAdapter.getCount() == 0) {
				showNoAudits(getView());
			}
		} catch (MobileClientException exc) {
			// Failed to delete audit
			showError(R.string.message_review_error_remove);
		}
	}

	/** Identifies logging source. */
	private static final String LOG_TAG = "MainMenuPage";

	/** Default e-mail recipient. */
	private static final String EMAIL_TO = "accounts@auditpro.io";

	/** References the list view. */
	private ListView auditList;

	/** References the sync menu item. */
	private MenuItem syncMenuItem;

	/* Completed audits fetched from the database. */
	List<Audit> completedAudits;

	/** Interfaces the completed audits to the list view. */
	private ArrayAdapter<Audit> auditAdapter;

	/** References the sync task in progress, if any. */
	private WeakReference<SyncTask> syncTask;
}

File: ios/Mobile Client/Mobile Client/pages/ReviewPage.swift
//
//  ReviewPage.swift
//  Mobile Client
//
//  Created by Eric Ruck on 4/4/18.
//  Copyright 2018-2019 AuditPro. All rights reserved.
//

import UIKit
import MessageUI


/**
 * Manages a fragment to display closed audits for review.
 * @author Eric Ruck
 */
class ReviewPage : BasePage, UITableViewDelegate, UITableViewDataSource, MFMailComposeViewControllerDelegate {
	@IBOutlet var noAuditsLabel: UILabel?
	@IBOutlet var auditTable: UITableView?

	/* Completed audits fetched from the database. */
	private var completedAudits: [Audit]? = nil

	/* Sync API client. */
	private var client: ApiClient?

	/* References title bar menu. */
	private var menuView: UIView? = nil

	/** Default e-mail recipient. */
	private let EMAIL_TO = ["accounts@auditpro.io"]
	private let EMAIL_SUBJ = "Completed Audit"
	private let EMAIL_BODY = "Completed audit of %@ attached"


	/**
	 * Initializes view contents on load.
	 */
	override func viewDidLoad() {
		super.viewDidLoad()

		// Connect to the audits database
		let db = AuditDatabase()
		completedAudits = db?.getCompleteAudits(userId: Security.userId)
		if (completedAudits?.count ?? 0 == 0) {
			showNoAudits()
		}
	}

	/**
	 * Provides the display page title.
	 * @return Display title
	 */
	override var pageName: String? {
		return "Review Closed Audits"
	}

	/**
	 * Makes sure we're not in the middle of an API call when we're
	 * exiting.
	 */
	override func onPageDisappearing() {
		client?.cancel()
	}

	/**
	 * Displays our menu in the status bar.
	 * @return Our menu view
	 */
	override func onCreateMenu() -> UIView? {
		// Have we loaded our menu view?
		if (menuView == nil) {
			// No, load it now
			menuView = Bundle.main.loadNibNamed("ReviewPageMenu",
				owner: nil, options: nil)![0] as? UIView
			if let syncAllView = menuView?.viewWithTag(1) as? UIButton {
				// Update the sync all option
				if (noAuditsLabel?.isHidden ?? false) {
					// Handle tap on sync all
					syncAllView.addTarget(self, action: #selector(onSyncAll), for: .touchUpInside)
				} else {
					// Nothing to sync
					syncAllView.isHidden = true
				}
			}
		}

		// Return the menu view
		return menuView
	}

	/**
	 * Shows the no audits state.
	 * @param view Our view
	 */
	private func showNoAudits() {
		noAuditsLabel?.isHidden = false
		auditTable?.isHidden = true
		menuView?.viewWithTag(1)?.isHidden = true
	}

	/**
	 * Gets the number of rows to display in the audit table.
	 * @param tableView View requesting count
	 * @param section Section query (only one section, should always be 0)
	 */
	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return completedAudits?.count ?? 0
	}

	/** Cached row height */
	private var rowHeight: CGFloat? = nil

	/**
	 * Provides the custom height for our cells.
	 * @param table Table in which cells live
	 * @param indexPath Path to row whose height we want
	 * @return Row height
	 */
	func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
		if (rowHeight == nil) {
			if let views = Bundle.main.loadNibNamed("ReviewPageTableCell", owner: nil, options: nil) {
				if let view = views[0] as? UITableViewCell {
					rowHeight = view.bounds.height
				}
			}
		}
		return rowHeight ?? 0
	}

	/**
	 * Provides a cell view for a table row.
	 * @param tableView Table who needs row view
	 * @param indexPath Path of needed row
	 * @return Row view
	 */
	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		// Get the cell to use
		var cell = tableView.dequeueReusableCell(withIdentifier: "audit")
		if (cell == nil) {
			cell = Bundle.main.loadNibNamed("ReviewPageTableCell", owner: nil, options: nil)![0] as? UITableViewCell
		}

		let storeLabel = cell?.viewWithTag(1) as! UILabel?
		let stampLabel = cell?.viewWithTag(2) as! UILabel?

		// Get the audit to show
		guard let audit = completedAudits?[indexPath.row] else {
			storeLabel?.text = "(Missing value)"
			stampLabel?.text = "(Missing value)"
			return cell!
		}

		// Populate the cell
		let formatter = DateFormatter()
		formatter.locale = Locale.current
		formatter.timeStyle = .none
		formatter.dateStyle = .medium
		storeLabel?.text = audit.storeDescr
		stampLabel?.text = formatter.string(from: audit.auditStartedAt)
		return cell!
	}

	/**
	 * Handles selection of a completed audit.
	 * @param tableView Table from which audit was selected
	 * @param indexPath Path of selected row
	 */
	func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		// Get the selected audit
		guard let audit = completedAudits?[indexPath.row] else {
			// Should never happen, no audit
			NSLog("Unexpected invalid row selection")
			return
		}

		// Clear the selection
		tableView.deselectRow(at: indexPath, animated: true)

		// Format audit description
		let formatter = DateFormatter()
		formatter.locale = Locale.current
		formatter.dateStyle = .short
		formatter.timeStyle = .none
		let message = String(format: "%@\nStarted at %@\nEnded at %@",
				audit.storeDescr,
				formatter.string(from: audit.auditStartedAt),
				formatter.string(from: audit.auditEndedAt!))

		// Show selected audit with options
		let alert = UIAlertController(title: message, message: nil, preferredStyle: .actionSheet)
		alert.addAction(UIAlertAction(title: "Sync", style: .default) { action in
			self.onSync(audit: audit)
		})
		if (MFMailComposeViewController.canSendMail()) {
			alert.addAction(UIAlertAction(title: "E-mail", style: .default) { action in
				self.onEmail(audit: audit)
			})
		}
		alert.addAction(UIAlertAction(title: "Edit", style: .default) { action in
			self.onEdit(audit: audit)
		})
		alert.addAction(UIAlertAction(title: "Remove", style: .default) { action in
			self.onRemove(audit: audit)
		})
		alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
		present(alert, animated: true)
	}


	/**
	 * Sets up an e-mail for the requested audit.
	 * @param audit Audit to e-mail
	 */
	private func onEmail(audit: Audit) {
		// Serialize the audit
		let db = AuditDatabase()
		guard
			let json = db?.serialize(audit: audit),
			let attachment = json.data(using: .utf8)
		else {
			// Failed to serialize the audit
			view.makeToast(message: "Error encoding audit attachment, please contact customer support.")
			return
		}

		// Format the attachment file name
		let formatter = DateFormatter()
		formatter.dateFormat = "'audit-'yyMMddHHmmssZ'.json'"
		let fileName = formatter.string(from: audit.auditStartedAt)

		// Send the e-mail
		Analytics.log(name: "Review", audit: audit, key: "Email", value: "Email")
		let mail = MFMailComposeViewController()
		mail.mailComposeDelegate = self
		mail.setToRecipients(EMAIL_TO)
		mail.setSubject(EMAIL_SUBJ)
		mail.setMessageBody(String(format: EMAIL_BODY, audit.description), isHTML: false)
		mail.addAttachmentData(attachment, mimeType: "application/json", fileName: fileName)
		present(mail, animated: true)
	}

	/**
	 * Handles e-mail composition completion.
	 * @param controller Composer controller
	 * @param result Completion result code
	 * @param error Error description or nil on success
	 */
	func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
		dismiss(animated: true)
	}

	/**
	 * Reopens the passed audit for editing.
	 * @param audit Audit to edit
	 */
	private func onEdit(audit: Audit) {
		// Attempt to reopen the audit
		guard let db = AuditDatabase() else {
			// Failed to access database
			view.makeToast(message: "Internal error, please contact customer support.")
			return
		}
		if db.resumeAudit(userId: Security.userId) != nil {
			// Another audit is currently in progress
			view.makeToast(message: "An audit is already in process. Complete the current audit from the main menu, then edit this one.")
			return
		}
		guard let reopened = db.reopenAudit(audit) else {
			// Failed to reopen this audit
			view.makeToast(message: "Unable to edit this audit at this time.")
			return
		}

		// Show the reopened audit
		Analytics.log(name: "Review", audit: audit, key: "Edit", value: "Reopen")
		host?.swapPage(SelectProductPage(audit: reopened))
	}

	/**
	 * Confirms request to remove the passed audit.
	 * @param audit Audit to remove
	 */
	private func onRemove(audit: Audit) {
		let message = String(format:
			"Are you certain you want to remove the audit %@? " +
			"It will be permanently deleted from your phone.",
			audit.description)
		let alert = UIAlertController(title: "Remove Audit",
			message: message, preferredStyle: .alert)
		alert.addAction(UIAlertAction(title: "Yes", style: .destructive) { action in
			Analytics.log(name: "Review", audit: audit, key: "Remove", value: "Remove")
			self.performRemove(audit: audit)
		})
		alert.addAction(UIAlertAction(title: "No", style: .cancel) { action in
			Analytics.log(name: "Review", audit: audit, key: "Remove", value: "Cancel")
		})
		present(alert, animated: true)
	}

	/**
	 * Removes the passed audit from the database.
	 * @param audit Audit to remove
	 */
	private func performRemove(audit: Audit) {
		// Remove the posted audit from our cache
		let db = AuditDatabase()
		guard db?.delete(audit: audit) ?? false else {
			view.makeToast(message: "Failed to remove audit, please contact customer support.")
			return
		}
		// Remove the audit from the list
		if let index = completedAudits?.firstIndex(of: audit) {
			completedAudits?.remove(at: index)
			auditTable?.reloadData()
		}
	}

	/**
	 * Confirms request to sync the passed audit.
	 * @param audit Audit to sync
	 */
	private func onSync(audit: Audit) {
		let message = String(format:
			"Are you certain you want to sync the audit %@? " +
			"On success the audit will be removed from your phone.",
			audit.description)
		let alert = UIAlertController(title: "Sync Audit",
			message: message, preferredStyle: .alert)
		alert.addAction(UIAlertAction(title: "Yes", style: .default) { action in
			Analytics.log(name: "Review", audit: audit, key: "Sync", value: "Sync")
			self.performSync(audit: audit)
		})
		alert.addAction(UIAlertAction(title: "Cancel", style: .cancel) { action in
			Analytics.log(name: "Review", audit: audit, key: "Sync", value: "Cancel")
		})
		present(alert, animated: true)
	}

	/**
	 * Performs the sync of a single audit.
	 * @param audit Audit to sync
	 */
	private func performSync(audit: Audit) {
		guard
			// Serialize the audit
			let db = AuditDatabase(),
			let json = db.serialize(audit: audit)
		else {
			// Failed to serialize the audit
			view.makeToast(message: "Failed to serialize the audit. Please contact customer support.")
			return
		}

		// Send the audit to the web service
		host?.isActivityShowing = true
		client = ApiClient(token: AppDelegate.sessionToken)
		client!.postPayload(auditJson: json, completion: { client in
			// Is the review page still current?
			self.client = nil
			if (self.host != nil) {
				DispatchQueue.main.async {
					// Update the user interface on the main thread
					self.host?.isActivityShowing = false
					if (client.message != nil) {
						self.view.makeToast(message: client.message!)
					} else {
						// Remove the posted audit from our cache
						db.delete(audit: audit)
						if let index = self.completedAudits?.firstIndex(of: audit) {
							self.completedAudits?.remove(at: index)
							self.auditTable?.reloadData()
							if (self.completedAudits?.count ?? 0) == 0 {
								self.showNoAudits()
							}
						}
					}
				}
			}
		})
	}

	/**
	 * Handles request to sync all waiting audits.
	 */
	@objc private func onSyncAll() {
		// Offer set unscanned options
		let message = AppDelegate.isWifiConnected
			? "Are you sure you want to sync all of the completed audits? Synced audits will be removed from yor phone."
			: "Are you sure you want to sync all of the completed audits? Synced audits will be removed from yor phone. Cellular data may be used."
		let alert = UIAlertController(title: "Sync All",
			message: message, preferredStyle: .alert)
		alert.addAction(UIAlertAction(title: "Yes", style: .default) { action in
			Analytics.menuAction(type: "Sync All", details: "sync")
			self.host?.requestSync()
		})
		alert.addAction(UIAlertAction(title: "Cancel", style: .cancel) { action in
			Analytics.menuAction(type: "Sync All", details: "cancel")
		})
		present(alert, animated: true)
	}
}
